### PR TITLE
Make allclose use isclose.

### PR DESCRIPTION
**Pull request by chebee7i**
Presently, `allclose` does not use `isclose`, preferring instead to use to `_allclose_points`, which barely differs from `isclose`. One of the subtle differences is that `isclose` has a bug that was previously fixed in #2280 for `allclose`. Similarly, `allclose` doesn't work nicely with `MaskedArray`, while `isclose` does. This PR fixes both issues. Finally, `allclose` doesn't have the option to compare NaNs as equal. This is now possible.

### CODE REVIEW

**Comment by charris** [comment]
Needs mention in the 1.10 release notes. Also, does the result change for any of the current uses of `allclose`?

**Comment by chebee7i** [comment]
I don't believe so. If it does, then it's a use case that didn't have a unit test, so it's tough to say. 

My reading of the code was that `allclose` properly handled `abs` on minimal values of integer types by casting, while `isclose` did not. So that was a bug for `isclose`. `allclose` did not handle NaN equality, while `isclose` did. Since we have `nan_equal=False`, it shouldn't change the behavior. Other than that, I think they were functionally equivalent (except for the `.all()` nature of `allclose`).

File: numpy/core/numeric.py [hunk 1]
```diff
@@ -2366,6 +2336,12 @@ def within_tol(x, y, atol, rtol):
 
     x = array(a, copy=False, subok=True, ndmin=1)
     y = array(b, copy=False, subok=True, ndmin=1)
+
+    # make sure y is an inexact type to avoid abs(MIN_INT); will cause
+    # casting of x later. Make sure to allow subclasses (e.g., for numpy.ma).
+    dtype = multiarray.result_type(y, 1.)
```
Comment (charris) [inline numpy/core/numeric.py]: Might be better to use a less suggestive variable name here, maybe `dt`.

File: numpy/testing/utils.py [hunk 2]
```diff
@@ -1289,7 +1291,8 @@ def assert_allclose(actual, desired, rtol=1e-7, atol=0,
     """
     import numpy as np
     def compare(x, y):
-        return np.core.numeric._allclose_points(x, y, rtol=rtol, atol=atol)
+        return np.core.numeric.isclose(x, y, rtol=rtol, atol=atol,
+                                             equal_nan=equal_nan)
```
Comment (charris) [inline numpy/testing/utils.py]: Should line up under the `x`.

**Comment by charris** [comment]
LGTM modulo a couple of style nitpicks.

File: numpy/core/numeric.py [hunk 3]
```diff
@@ -2262,6 +2227,9 @@ def allclose(a, b, rtol=1.e-5, atol=1.e-8):
         The relative tolerance parameter (see Notes).
     atol : float
         The absolute tolerance parameter (see Notes).
+    equal_nan : bool
+        Whether to compare NaN's as equal.  If True, NaN's in `a` will be
+        considered equal to NaN's in `b` in the output array.
```
Comment (charris) [inline numpy/core/numeric.py]: Need a `.. versionadded:: 1.10` here, either before the first line (I'm tending that way), of after a blank line at the end.

Comment (chebee7i) [inline numpy/core/numeric.py]: What is the appropriate indentation level?  With the docstring, or aligned with "Whether". And should I have a blank line after the versionadded line?

Comment (charris) [inline numpy/core/numeric.py]: Same as the text. You can see plenty of examples if you `grep -r versionadded numpy`.

Comment (chebee7i) [inline numpy/core/numeric.py]: Well, that is odd....the change appears in e7b3a0c, but not under "Files changed" for this PR.

**Comment by chebee7i** [comment]
Fixed (and force pushed).

**Comment by charris** [comment]
Thanks @chebee7i .

**Comment by taldcroft** [comment]
FYI this change introduced a large number of test errors in https://github.com/astropy/astropy/issues/3584.

**Comment by mhvk** [comment]
@taldcroft is correct, though I would add that the errors came about because subclasses are now passed on correctly (rather than turned into plain arrays), which in general I think is a *good thing*. Even if it was a pain to adapt to many of them, our tests are now better, so overall good that this change was made.

**Pull request merged**